### DISCUSSION TITLE
LDM Notes for September 11th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8424] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-09-11.md

### Agenda

- Better conversion from collection expression and `params` collections
- First-class span types
    - `Reverse`
    - New ambiguities
    - Covariant arrays

**Comment by huoyaoyuan** [comment]
I was hitting `Reverse` with covariant in https://github.com/dotnet/runtime/pull/105407 , but not overload resolution case.

In-place `Reverse` as well as `Shuffle` are quite special operations. They require writable destination, but are tolerate with covariance because all written items come from the original storage. I'm not sure whether it's actionable to make improvements here.

**Comment by KennethHoff** [comment]
( on betterness changes):

> we want to make the rules for the new feature consistent with the **new** rules for the old feature.

Is this a typo? Shouldn't it be "consistent with the rules for the old feature." (i.e. not "new rules for old feature")

**Reply by 333fred** [reply]
Nope. The new rules for the old feature is params collections. Ie, all 3 (collection expressions, params arrays, params collections) are now as consistent as we can can make them.

**Reply by KennethHoff** [reply]
but params collections aren't an old feature? 

Params arrays are old, are they changing too?

Collections collections are arguably old - and they are changing (?) - is this the argument?

**Reply by 333fred** [reply]
> but params collections aren't an old feature?

They're new rules on an old feature.

**Reply by KennethHoff** [reply]
Ah, so params arrays and the new params collections are the same feature in the LDM's eyes? Fair enough 👍🏻

**Comment by tfenise** [comment]
I would argue the conversion `T[] -> Span<T>` should not be automatically preferred over `T[] -> IEnumerable<T>`(or more generally, `IReadOnly*<T>`) at all, because a method with a `Span<T>` parameter may modify the span (otherwise it should just accept `ReadOnlySpan<T>`), but a method with an `IEnumerable<T>` parameter probably never modifies the collection, so the semantics of such two methods are probably different even if they have the same name, and one of them should not be automatically preferred over the other.

I am also curious why a rule of "Better conversion from expression" is proposed to be [introduced](https://github.com/dotnet/csharplang/blob/7bfaafb048c89e6a8998938203e3ef24c8497cdc/proposals/first-class-span-types.md#better-conversion-from-expression), instead of a rule of "[Better conversion target](https://github.com/dotnet/csharpstandard/blob/8c5e008e2fd6057e1bbe802a99f6ce93e5c29f64/standard/expressions.md#12647-better-conversion-target)" to prefer `(ReadOnly)Span<T>` over `I(ReadOnly)*<T>`, etc. Unconditionally preferring implicit span conversions over implicit conversions that are not implicit span conversions sounds inappropriate. Why should `T[] -> Span<T>` be preferred over `T[] -> ArraySegment<T>`? `T[] -> ArraySegment<T>` sounds better.

**Reply by 333fred** [reply]
> I would argue the conversion T[] -> Span<T> should not be automatically preferred over T[] -> IEnumerable<T>(or more generally, IReadOnly*<T>) at all, because a method with a Span<T> parameter may modify the span (otherwise it should just accept ReadOnlySpan<T>), but a method with an IEnumerable<T> parameter probably never modifies the collection, so the semantics of such two methods are probably different even if they have the same name, and one of them should not be automatically preferred over the other.

This is just how more specific types work in general. We're trying to emulate `ReadOnlySpan<T>` implementing `IEnumerable<T>`, and `Span<T>` inheriting from `ReadOnlySpan<T>`. This makes `Span<T>` a more specific type than `IEnumerable<T>`; to take an existing analogy, between `IEnumerable<T>` and `IList<T>`, the latter is preferred for an array, because it is more specific, as shown in [this sharplab example](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbjzwDcBDMIhsOgTyIF4ioBTAdyIBLKMADaAXSIBvIphhF0RAL5VcAYQB0AWQAUzNgEpVeVAGYSitdLxFbJc6QBsJMkS0AeACoA+HaVNe3kQ8BpxBpACcOgBEAKJQEAC2PCwARgA2PNFGNnZmJJjOqK4ePjoAkgAyggDOwIHBoRzhmFHR1XXZqkpAA===).

> Why should T[] -> Span<T> be preferred over T[] -> ArraySegment<T>? T[] -> ArraySegment<T> sounds better.

Because both the BCL and C# language design team feel that this is the direction we want the language to move. `(ReadOnly)Span` is the type we prefer here; `ArraySegment` was never heavily adopted in the BCL, and for many years now (since C# 7) the team has been moving in this direction.

**Reply by tfenise** [reply]
Following the same logic of how more specific types work in general, `T[] -> ArraySegment<T>` should be preferred over `T[] -> Span<T>`. Indeed, there is an user-defined implicit conversion from `ArraySegment<T>` to `Span<T>` in the BCL.

This would also be a breaking change, as currently, `T[] -> ArraySegment<T>` is preferred over `T[] -> Span<T>`.

**Reply by 333fred** [reply]
Well, sure, but we don't really have APIs overloaded on `ArraySegment<T>` and `Span<T>`; as I mentioned, the BCL and language teams do not feel that `ArraySegment<T>` should be used over `Span<T>`, and API design has been moving towards `(ReadOnly)Span`s for more than half a decade.

**Reply by tfenise** [reply]
> we don't really have APIs overloaded on `ArraySegment<T>` and `Span<T>`;

Third-party libraries may have APIs overloaded on `ArraySegment<T>` and `Span<T>`.

> API design has been moving towards `(ReadOnly)Span`s for more than half a decade.

API design and overload resolution are different issues.

If the logic of "how more specific types work in general" as a general principle is to be honoured, I don't see what is so special about `T[] -> (ReadOnly)Span<T>` that they should be introduced as a special case and always preferred over any other implicit conversions.

**Reply by 333fred** [reply]
> API design and overload resolution are different issues.

One influences the other, you absolutely cannot separate these two things.

**Reply by tfenise** [reply]
Yes, but `Span<T>` should be preferred over `ArraySegment<T>` in the sense of API design does not mean `Span<T>` should be preferred over `ArraySegment<T>` in the sense of overload resolution.

**Reply by 333fred** [reply]
It absolutely does, because overload resolution impacts API design.

**Reply by tfenise** [reply]
From the perspective of `ref struct` implementing interfaces, a new rule of [Better conversion target](https://github.com/dotnet/csharpstandard/blob/8c5e008e2fd6057e1bbe802a99f6ce93e5c29f64/standard/expressions.md#12647-better-conversion-target) could be introduced by imposing a `ref struct` be a better conversion target than the interfaces that it implements. This rule would resolve the ambiguity between `T[] -> ReadOnlySpan<T>` and `T[] -> IEnumerable<T>` without introducing the ad hoc and suspicious rule that implicit span conversions be always preferred in any case.

**Reply by timcassell** [reply]
It wouldn't, because `ReadOnlySpan<T>` cannot implement `IEnumerable<T>`.

**Reply by tfenise** [reply]
OK then, but the rule that implicit span conversions be always preferred in any case is still suspicious. Again, `Span<T>` should be preferred over `ArraySegment<T>` in the sense of API design does not mean `Span<T>` should be preferred over `ArraySegment<T>` in the sense of overload resolution. For example, `Type` should be preferred over `TypeInfo` in the sense of API design (https://github.com/dotnet/runtime/issues/53217), but this does not mean we should introduce a new rule of overload resolution that prefers implicit conversions to `Type`, the base class.

**Reply by 333fred** [reply]
> Again, Span<T> should be preferred over ArraySegment<T> in the sense of API design does not mean Span<T> should be preferred over ArraySegment<T> in the sense of overload resolution.

I still disagree with this statement. API design absolutely needs to take overload resolution into account, and I've said several times, the future of .NET is in Spans, not ArraySegments.

**Reply by tfenise** [reply]
There is another problem with the proposed rule that implicit span conversions be always preferred in any case, which is that it does not resolve the ambiguity between `ArraySegment<T> -> Span<T>` and `ArraySegment<T> -> IList<T>`. `ArraySegment<T> -> Span<T>` is just a user-defined implicit conversion, not an implicit span conversion provided by the language, so the proposed rule does not help. Instead, the potential rule that `Span<T>` be better conversion target than `IList<T>` would cover this case.

I am not saying that the whole first-class Span types proposal should take user-defined types like `ArraySegment<T>` into account. That would be too complicated. However, the potential rule that `(ReadOnly)Span<T>` be better conversion target than `I(ReadOnly)*<T>` is no more complicated than the proposed rule that implicit span conversions be always preferred in any case, and resolves more ambiguities, so it is more desirable.

**Reply by tfenise** [reply]
There is even another problem with the proposed rule that implicit span conversions be always preferred in any case, which is that it prefers `DerivedClass[] -> ReadOnlySpan<BaseClass>` over `DerivedClass[] -> IEnumerable<DerivedClass>`. Arguably, there should be an ambiguity between `DerivedClass[] -> ReadOnlySpan<BaseClass>` and `DerivedClass[] -> IEnumerable<DerivedClass>`, as there is an ambiguity between `DerivedClass[] -> IReadOnlyList<BaseClass>` and `DerivedClass[] -> IEnumerable<DerivedClass>`.

**Reply by CyrusNajmabadi** [reply]
These don't seem like problems to me. We discussed what should happen in cases like this and were happy with the direction.

**Reply by jaredpar** [reply]
> I've said several times, the future of .NET is in Spans, not ArraySegments.

The present of .NET is Spans. 😄
